service/grails-app/services/org/olf/rs/PatronNoticeService.groovy

```diff
@@ -0,0 +1,132 @@
+package org.olf.rs
+
+import com.k_int.okapi.OkapiClient
+import grails.core.GrailsApplication
+import grails.gorm.multitenancy.Tenants
+import groovy.json.JsonSlurper
+import org.apache.kafka.clients.consumer.KafkaConsumer
+
+import org.olf.rs.EmailService
+import org.olf.rs.EventConsumerService
+import org.olf.rs.EventPublicationService
+import org.olf.rs.NoticePolicyNotice
+
+public class PatronNoticeService {
+
+  EmailService emailService
+  EventConsumerService eventConsumerService
+  EventPublicationService eventPublicationService
+  OkapiClient okapiClient
+  GrailsApplication grailsApplication
+  private KafkaConsumer consumer = null
+
+  @javax.annotation.PostConstruct
+  public void init() {
+    // We need another consumer because EventConsumerService consumes all topics
+    // as they come in whereas the notice triggers need to queue until they can
+    // be consumed under the aegis of a timer request
+    log.debug("Configuring event consumer for patron notice service")
+    Properties props = new Properties()
+    try {
+      grailsApplication.config.events.consumer.toProperties().each { final String key, final String value ->
+        // Directly access each entry to cause lookup from env
+        String prop = grailsApplication.config.getProperty("events.consumer.${key}")
+        log.debug("Configuring event consumer for patron notice service :: key:${key} value:${value} prop:${prop}");
+        props.setProperty(key, prop)
+      }
+      log.debug("Configure consumer ${props}")
+    }
+    catch ( Exception e ) {
+      log.error("Problem assembling props for consume",e);
+    }
+
+    consumer = new KafkaConsumer(props)
+    log.debug("PatronNoticeService::init() returning");
+  }
+
+  @javax.annotation.PreDestroy
+  private void cleanUp() throws Exception {
+    log.info("PatronNoticeService::cleanUp");
+    consumer.wakeup();
+  }
+
+  public void triggerNotices(PatronRequest pr, String trigger) {
+    if (!(pr?.patronEmail instanceof String)) {
+      log.error("Notice triggered for request ${pr.hrid} with no email for patron");
+      return;
+    }
+    log.debug("triggerNotices(${pr.patronEmail}, ${trigger})")
+    String tenant = Tenants.currentId()
+    eventPublicationService.publishAsJSON(
+      "${tenant}_PatronNoticeEvents",
+      null,
+      [
+        event:'TriggerNotices_ind',
+        tenant: tenant,
+        payload:[
+          trigger: trigger,
+          email: pr.patronEmail,
+          values: [
+            user: [
+              id: pr?.patronReference,
+              firstName: pr?.patronGivenName,
+              lastName: pr?.patronSurname,
+            ],
+            request: [
+              id: pr.hrid,
+              // pickupLocation: pr?.pickupLocation,
```

Comment:
The safe navigation operator doesn't seem to be sufficient here eg. it doesn't just happily set this to null. What's a nice Groovy way to add this property if available and not if not (though iirc we may actually require both of these things on the record so all valid records will have them populated so it could legitimately be an exception were they not present)

Reply:
As on Slack, I recommend
```
Map map = [
//some stuff here
]
if (a) {
  map["a"] = a
}

```